[MRG + 2] EnsembleClassifier implementation

rasbt opened the pull request:
This is an implementation of an estimator to combine scikit-learn classification estimators into an ensemble classifier for majority rule class label prediction or classification based on weighted probability averages.

I added a section at the bottom of the ensemble.rst documentation with a more detailed description and examples.

I'd be happy to make improvements or adjustments if necessary. Please let me know, I am happy to help!

rasbt committed:
ensemble classifier 1st commit

jnothman commented:
I think EnsembleClassifier is probably a bit too general. Perhaps `VotedClassifier` or `VotingClassifier` or `MajorityRuleClassifier`.

rasbt commented:
I agree, Joel. I like MajorityRuleClassifier, it's pretty much summarizes the concept in one phrase.

jnothman commented:
except that often you will be relying on the highest vote, not a strict
majority

On 26 January 2015 at 16:57, Sebastian Raschka <notifications@github.com>
wrote:

>  I agree, Joel. I like MajorityRuleClassifier, it's pretty much
> summarizes the concept in one phrase.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4161#issuecomment-71417634>
> .
>

jnothman commented:
Not that I've read the exact algorithm.

On 26 January 2015 at 19:39, Joel Nothman <joel.nothman@sydney.edu.au>
wrote:

> except that often you will be relying on the highest vote, not a strict
> majority
>
> On 26 January 2015 at 16:57, Sebastian Raschka <notifications@github.com>
> wrote:
>
>>  I agree, Joel. I like MajorityRuleClassifier, it's pretty much
>> summarizes the concept in one phrase.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/4161#issuecomment-71417634>
>> .
>>
>
>

ragv commented on sklearn/ensemble/majorityrule_classifier.py:
```diff
@@ -0,0 +1,166 @@
+        """
+        if self.weights:
+            avg = self.predict_proba(X)
+            maj = np.apply_along_axis(lambda x: max(enumerate(x), key=operator.itemgetter(1))[0], axis=1, arr=avg)
```
Could you refactor this to two lines maybe? 
( PEP8 specifications say that max line length should be 79... )

ragv commented on sklearn/ensemble/majorityrule_classifier.py:
```diff
@@ -0,0 +1,166 @@
+        
+        else:
+            self.classes_ = self._get_classes(X)
+            maj = np.asarray([np.argmax(np.bincount(self.classes_[:,c])) for c in range(self.classes_.shape[1])]) 
```
Same here.... :)

glouppe commented:
The algorithm you have implemented is actually called "soft voting" rather than "majority voting". In the former, the prediction is the argmax of the sums of predict probas, in the latter the prediction is the argmax of the sums of predictions. 

By the way, a more general algorithm is "stacking", which basically learns a new estimator from the predictions of others. Shouldn't we try to go in that direction instead, with majority voting/soft voting as special cases?

arjoly commented:
> By the way, a more general algorithm is "stacking", which basically learns a new estimator from the predictions of others. Shouldn't we try to go in that direction instead, with majority voting/soft voting as special cases?

+1

amueller commented:
It seemed to me that stacking was less used in practice, but Owen (from data bricks, Kaggle #1) said he uses it quite a bit.
I think it would be nice to have stacking but the simple case is also helpful, right? So why not implement this first, and possibly add stacking later? The implementation is pretty simple anyhow...

rasbt commented:
> The algorithm you have implemented is actually called "soft voting" rather than "majority voting". In the former, the prediction is the argmax of the sums of predict probas, in the latter the prediction is the argmax of the sums of predictions.

Hm, sounds like that implemented both at once then. If weights are provided, it is a "soft voting" (based on probas) and without the weights "majority voting" (based on class labels)

Do you think it would make sense to split the current implementation into a SoftVotingClassifier and a MajorityRuleClassifier for clarity?

About "stacking": I agree with @aemuller that this could be implemented additionally later. E.g., by using the attributes from the MajorityRuleClassifier as input for a new classifier. On the other hand, would this need an implementation? Maybe an example would be sufficient that shows how the returned predictions/probas from the MajorityRule/SoftVoting classifier can be used as input for a new classifier...

amueller commented:
@rasbt in stacking (see ESL) the probabilities of the first stage of classifiers are obtained using out of sample estimates using cross-validation, so it is a bit more complex.

amueller commented:
I wouldn't split this up as two classes, but I would add a different string options for using majority or soft voting. Maybe call it ``VotingClassifier`` and add a ``voting`` parameter that can be ``hard``, ``soft`` or a sequence of weights? Not sure if the name ``voting`` is very discoverable, ``weights`` seems more natural.

amueller commented:
With you current interface it would also be possible to just add a ``uniform`` keyword, which might be simpler than specifying a constant vector.

rasbt commented:
@amueller Good idea! I added the `voting` parameter which can be set to `'soft'` or `'hard'`. 

About the `uniform` keyword: I don't think it is necessary. Right now, uniform weights are used by default if `weights=None`, which is maybe cleaner because `np.average` also has a `weights` parameter.
`avg = np.average(self.probas_, axis=0, weights=self.weights)`
Or would you think that an explicit `'uniform'` keyword could be beneficial for clarity?

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,195 @@
+        self.voting = voting
+        
+        if weights and len(weights) != len(clfs):
+            raise ValueError('Number of classifiers and weights must be equal')      
```
maybe also give the numbers here.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,195 @@
+
+    def fit(self, X, y):
+        """ 
+        Fits the scikit-learn estimators.
```
I think scikit-learn inside scikit-learn is redundant ;)

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,195 @@
+        else:
+            return self._get_classes(X)  
+    
+    def _get_classes(self, X):
```
this should be ``get_prediction`` or ``get_predict`` or just ``_predict``. The ``classes_`` attribute is ``np.unique(y)``

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,195 @@
+            maj = np.apply_along_axis(lambda x: max(enumerate(x), key=operator.itemgetter(1))[0], axis=1, arr=avg)
+        
+        else: # 'hard' voting
+            self.classes_ = self._get_classes(X)
```
The ``classes_`` attribute should be ``np.unique(y)``, not this.

amueller commented:
I don't see where the weights are used in the prediction currently. If you let the user specify both the weights and the voting, then you need to make sure that the user doesn't shoot herself in the foot easily. So if ``voting='hard'`` and some weights are specified, they should also be used, I guess for weighted voting.

I don't think you need to add an additional keyword ``uniform``. It is just that a user should be able to use both, hard and soft voting, without having to specify the number of classes in the estimator construction. I think that is possible with your current solution.

Something that might be tricky is that ``predict_proba`` should be consistent with ``predict`` in all cases. For hard voting that causes some minor issues, see #3891. You could just return the normalized votes but I am not a fan of that, as you throw away so much information (and your roc score will be abysmal).

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,195 @@
+    Parameters
+    ----------
+    clfs : array-like, shape = [n_classifiers]
+      A list of scikit-learn classifier objects.
```
Same comment about redundancy.

amueller commented on sklearn/ensemble/__init__.py:
```diff
@@ -28,5 +29,6 @@
            "BaggingRegressor", "GradientBoostingClassifier",
            "GradientBoostingRegressor", "AdaBoostClassifier",
-           "AdaBoostRegressor", "bagging", "forest", "gradient_boosting",
+          "AdaBoostRegressor", "VotingClassifier"
```
indentation?

amueller commented:
You should add the class in ``sklearn.utils.testing`` to ``DONT_TEST`` I think.

rasbt commented:
@amueller Thanks for the comments, I think/hope I've addressed all of your points except for the "predict_proba should be consistent" part.

> I don't see where the weights are used in the prediction currently. If you let the user specify both the weights and the voting, then you need to make sure that the user doesn't shoot herself in the foot easily. So if voting='hard' and some weights are specified, they should also be used, I guess for weighted voting.

The weights are being used now for the "hard" voting in the most recent update. I didn't implement it earlier because I really wouldn't want to use something like this but prefer the weighted avg. probabilities instead. But you are right about the "shooting in the foot" ;)

> Something that might be tricky is that predict_proba should be consistent with predict in all cases. For hard voting that causes some minor issues, see #3891. You could just return the normalized votes but I am not a fan of that, as you throw away so much information (and your roc score will be abysmal).

Here, the challenge was that not all classifiers have the `predict_proba` (I think SVMs?); doing it like this way just allows the usage of "hard" voting if such classifiers are being used in the "ensemble."

amueller commented:
Some tests and doctests are missing imports, see travis.

amueller commented:
Have you checked what happens in the multi-label case? If you don't want to implement that for the moment, please make sure an appropriate error message is raised.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,202 @@
+      
+        Returns
+        -------
+        If not `weights=None`:
```
this is not correct any more, right?

amueller replied:
Also, I don't think this is valid numpy doc format ;)
Lastly, I think the you should flatten the array to be ``(n_samples, *)`` and not have ndim=3, so that can directly be used in a pipeline.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,202 @@
+        """
+        if self.voting == 'soft':
+            avg = self.predict_proba(X)
+            maj = np.apply_along_axis(lambda x: max(enumerate(x), key=operator.itemgetter(1))[0], axis=1, arr=avg)
```
why is this not simply ``np.argmax(avg, axis=1)``?

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,202 @@
+            maj = np.apply_along_axis(lambda x: max(enumerate(x), key=operator.itemgetter(1))[0], axis=1, arr=avg)
+        
+        else: # 'hard' voting
+            self.classes_ = self._predict(X)
```
As I said in an outdated diff, ``self.classes_ == np.unique(y_train)``.

amueller commented:
your implementation currently assumes ``np.unique(y) == np.arange(n_classes)`` as far as I can see. You need a ``LabelEncoder`` or use ``self.classes_, y = np.unique(y, return_inverse)`` or something like that.

mblondel commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,202 @@
+      
+    """
+    def __init__(self, clfs, voting='hard', weights=None):
+        self.clfs = clfs
```
Use `estimators` instead of `clfs` for consistency with the rest of the scikit.

mblondel commented:
> I wouldn't split this up as two classes, but I would add a different string options for using majority or soft voting. Maybe call it VotingClassifier and add a voting parameter that can be hard, soft

+1, that was my feeling too. Stacking can go to another class.

The class could also accept classifiers which are already fitted. This would be useful for find the best parameters of each classifier independently. Naively using GridSearchCV on VotingClassifier would try all hyper-parameter combinations, even though classifiers are independent.

> but Owen (from data bricks, Kaggle #1)

Datarobot :)

rasbt commented:
Hope it looks better this time.... Meanwhile thanks for the patience!
Sorry about the tests: Added the doc tests now and tested it via `python sklearn/ensemble/tests/test_voting_classifier.py` I am a little bit confused by now: I tested it before by running `nosetests` in the `sklearn/ensemble/tests/` directory and everything seemed to be fine -- I am wondering why it didn't pick up the problems with the incorrect imports...

Good call with the LabelEncoder btw. I didn't notice/think about it previously

rasbt commented:
Regarding the multi-label classification: I've never really done this and I am really unfamiliar with implementations in scikit-learn. I doubt that it would work with this ensemble classifier implementation, thus, an error message would only make sense.  Do you have any good recommendation how to catch if a user is using an estimator for multi-label classification?

glouppe commented:
> I think it would be nice to have stacking but the simple case is also helpful, right? So why not implement this first, and possibly add stacking later? The implementation is pretty simple anyhow... 

Yes sure, having a simple voting classifier is already quite helpful :)

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+        
+        if voting not in ('soft', 'hard'):
+            raise ValueError("voting must be 'soft' or 'hard'; got (voting=%r)"
+                             % voting)
```
The validation of parameters should happen at `fit`, not at initialization.

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+        self.voting = voting
+        
+        if weights and len(weights) != len(estimators):
+            raise ValueError('Number of classifiers and weights must be equal')      
```
Same here

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+        
+        self.weights = weights
+        self.le = LabelEncoder()
+        self.classes_ = None
```
In the same spirit, learned attributes should be created at `fit`.

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+            Weighted average probability for each class per sample.
+        
+        """
+        self.probas_ = self._predict_probas(X)
```
Probas should not be saved as attributes.

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+            Class labels predicted by each classifier.
+        
+        """
+        if self.weights:
```
This should be controlled by `self.voting` instead

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+            Predicted class labels by majority rule.
+        
+        """
+        if self.voting == 'soft':
```
Shouldnt we take weights into account as well? even for soft voting?

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,204 @@
+            maj = self.le.inverse_transform(np.argmax(avg, axis=1))
+        
+        else: # 'hard' voting
+            self.classes_ = self._predict(X)
```
`self.classes_` should not be changed during `predict`

amueller commented:
On 01/26/2015 09:16 PM, Mathieu Blondel wrote:
>
>     but Owen (from data bricks, Kaggle #1
>     <https://github.com/scikit-learn/scikit-learn/issues/1>)
>
> Datarobot :)
>
brainfart, sorry

amueller commented:
For the testing: you should run ``nosetests -sv sklearn/ensemble/tests``  for the tests, and it should pick up missing imports in the test files, but not in the doctests. For the doctests, use ``nosetests -sv --with-doctest  sklearn/ensemble`` I think.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+        self.estimators = estimators
+        self.voting = voting
+        self.weights = weights
+        self.le = LabelEncoder()
```
should be constructed in ``fit``

amueller replied:
And also have a trailing underscore, ``self.le_`` (and maybe check what they are called in other modules, see if the shorthand le is used).

rasbt replied:
Looks like both `lb` : sklearn/metrics/classification.py 

and `le`: sklearn/neighbors/nearest_centroid.py, sklearn/neighbors/nearest_centroid.py , etc.

are being used. I would go with `le` (majority voting rule ;)).

Maybe the `lb` in sklearn/metrics/classification.py  should be adjusted to `le` for concistency?

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+            clf.fit(X, y)
+        
+        self.le.fit(y)
+        self.classes_ = np.unique(y)
```
I think if the estimator has a label encoder, we use
``python
@property
def classes_(self):
    return self.le_.classes_
``
but please double check.

rasbt replied:
Sorry, couldn't find anything like that. But it should be the same as `self.classes_ = np.unique(y)`, right?

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+        return np.asarray([clf.predict_proba(X) for clf in self.estimators])
+        
+
+if __name__ == "__main__":
```
This should go.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+            maj = self.le.inverse_transform(np.argmax(avg, axis=1))
+        
+        else: # 'hard' voting
+            cl_lab = self._predict(X)
```
please use as little shorthand as possible, maybe just use ``predictions`` here or something (or ``per_classifier_pred``)

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+        else: # 'hard' voting
+            cl_lab = self._predict(X)
+            
+            # arange class labels as [n_cl_lab, n_classifier_predictions]
```
what is are these two numbers? is the second one ``n_samples``?

rasbt replied:
I couldn't find a meaningful description for it. E.g., if 3 classifiers are used in the VotingClassifier, this would look like
    
array([[0, 0, 0],
           [0, 0, 0],
          ...,
           [2, 2, 2]])

where the 1st column are the class labels predicted by the 1st classifier, the 2nd column the class labels predicted by the 2nd classifier etc.

Maybe something like `[n_classifier_results, n_classifier]` is more meaningful?

amueller replied:
I'd say ``[n_samples, n_classifiers]``

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+        
+        Returns
+        ----------
+        maj : array-like, shape = [n_cl_lab]
```
is the shape not ``[n_samples]``?

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+            cl_lab = self._predict(X)
+            
+            # arange class labels as [n_cl_lab, n_classifier_predictions]
+            cl_lab = np.asarray([cl_lab[:,c] 
```
What does this line do? Was that not the shape before?

rasbt replied:
It's basically converting it from `[n_classifiers, n_samples]` to `[n_samples, n_classifiers]` 

E.g.,

    cl_lab = self._predict(X)

    array([[0, 0, 0, 0, 1, 1, 2, 2, 2, 2],
       [0, 0, 0, 0, 1, 1, 1, 2, 2, 2],
       [0, 0, 0, 0, 1, 1, 1, 2, 2, 2]])

    cl_lab = np.asarray([cl_lab[:,c] 
    array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0],
       [0, 0, 0],
       [1, 1, 1],
       [1, 1, 1],
       [2, 1, 1],
       [2, 2, 2],
       [2, 2, 2],
       [2, 2, 2]])

amueller replied:
Right. Why don't you use ``.T``? Also, maybe just return ``.T`` from ``_predict``. Most things in sklearn have ``n_samples`` as the first dimension.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,206 @@
+                        for c in range(cl_lab.shape[1])])
+            
+            # dupl. class labels if weights are provided for argmax maj. rule
+            if self.weights:
```
can't you just use the ``weights`` parameter of ``bincount``? When was that introduced?

rasbt replied:
Great! For some reason I didn't know that there was a `weights` parameter in `bincount` -- This way it's much cleaner and more efficient. That will also come in handy for some of my other projects, thanks!

rasbt commented:
That's very helpful information @amueller , and also thanks for the suggestions @glouppe .
This has been quite a good learning experience so far! 
The tests seem to pass now, I hope I didn't leave any mess "style-wise"

amueller commented:
No worries. We are a bit strict around these parts ;) I haven't checked tests or documentation yet but this PR is definitely heading in a good direction!

rasbt commented:
It's a very, very good thing that you are strict about those things. I particularly like scikit-learn for its design, which makes it easy to understand and use.

I just saw your comments on the diffs above.I will get to this soon, so please don't spend your time checking those things just yet ;)

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,198 @@
+        
+        self.le_ = LabelEncoder()
+        self.le_.fit(y)
+        self.classes_ = np.unique(y)
```
oh yeah, maybe we abandoned that pattern. You can use ``self.le_.classes_`` though.

rasbt commented:
Just checked back and see the Travis check passed now. I added all the suggested changes (and I hope it didn't miss anything)...
@amueller Whoa, this transpose thing was kind of embarrassing: One of those "cannot see the wood for the trees" moments, thanks!

amueller commented:
Happens to all of us ;) I'll have a look later.

rasbt commented:
Sorry for bothering, looks like there is a lot of stuff in terms of pull requests and issues to be processed, did you have a chance to look at the most recent update, yet? 

I was just wondering if something like a "changelog" needs to be updated if you want to pull this feature?

glouppe commented on doc/modules/ensemble.rst:
```diff
@@ -892,3 +892,95 @@ averaged.
+========================
+
+The idea behind the voting classifier implementation is to combine conceptually different machine learning classifiers and use a majority vote or the average predicted probabilities (soft vote) to predict the class labels. Such a classifier can be useful for a set of equally well performing model in order to balance out their individual weaknesses.
+
```
Can you break this line into several lines, each no longer than 80 chars?

glouppe commented on doc/modules/ensemble.rst:
```diff
@@ -892,3 +892,95 @@ averaged.
+Majority Class Labels (Majority/Hard Voting) 
+--------------------------------------------
+
+In majority voting, the predicted class label for a particular sample is the class label that represents the majority (mode) of the class labels predicted by each individual classifier. 
```
Same here

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+    voting : str, {'hard', 'soft'} (default='hard')
+      If 'hard', uses predicted class labels for majority rule voting.
+      Else if 'soft', predicts the class label based on the argmax of 
+      the sums of predict probalities.
```
predicted probabilities

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+    weights : array-like, shape = [n_classifiers], optional (default=`None`)
+      Sequence of weights (`float` or `int`) that are multiplied with the 
+      predicted class probabilities before averaging if `voting='soft'` 
+      Uses uniform weights if `None`.
```
Weights are also used in hard voting, which is not stated here.

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+    def _predict_probas(self, X):
+        """ Collects results from clf.predict calls. """
+        return np.asarray([clf.predict_proba(X) for clf in self.estimators])
+        
```
Add an empty line here

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+        self.classes_ = self.le_.classes_
+ 
+        for clf in self.estimators:
+            clf.fit(X, y)
```
Each classifier should be fit on `self.le_.transform(y)` instead of `y`.

glouppe commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+        Returns
+        ----------
+        maj : array-like, shape = [n_samples]
+            Predicted class labels by majority rule.
```
remove "by majority rule"

glouppe commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -0,0 +1,49 @@
+    scores = cross_validation.cross_val_score(eclf, X, y, cv=5, scoring='accuracy')
+    assert_almost_equal(scores.mean(), 0.93, decimal=2)   
+
+
```
In general, could you write tests checking for the correctness of the results? (E.g., by comparing the vectors returned by `predict` or `predict_proba` with the ones that you would manually obtain on a toy problem?)

glouppe commented:
Regarding style conventions, could you also make sure that everything is pep8 and pep257 compliant?

glouppe commented:
What is the coverage of the current tests by the way? Make sure everything is tested.

amueller commented on doc/modules/ensemble.rst:
```diff
@@ -892,3 +892,95 @@ averaged.
+- classifier 2 -> class 1
+- classifier 3 -> class 2
+
+the VotingClassifier (with ``voting='hard'``) would classify the sample as "class 1" based on the majority class label.
```
What happens for ties?

amueller commented on doc/modules/ensemble.rst:
```diff
@@ -892,3 +892,95 @@ averaged.
+    >>> from sklearn.ensemble import VotingClassifier
+    >>> import numpy as np
+
+    >>> np.random.seed(123)
```
It would probably be preferred to just set the seed everywhere so that the test is side-effect free.

amueller commented on doc/modules/ensemble.rst:
```diff
@@ -892,3 +892,95 @@ averaged.
+In contrast to majority voting (hard voting), soft voting returns the class label as argmax of the sum of predicted probabilities.
+
+
+Specific weights can be assigned to each classifier via the ``weights`` parameter. When weights are provided, the predicted class probabilities for each classifier are collected, multiplied by the classifier weight, and averaged. The final class label is then derived from the class label with the highest average probability.
```
The same applies to the votes above.

amueller commented on doc/modules/ensemble.rst:
```diff
@@ -892,3 +892,95 @@ averaged.
+
+Specific weights can be assigned to each classifier via the ``weights`` parameter. When weights are provided, the predicted class probabilities for each classifier are collected, multiplied by the classifier weight, and averaged. The final class label is then derived from the class label with the highest average probability.
+
+To illustrate this with a simple example, let's assume we have 3 classifiers and a 3-class classification problems where we assign equal weights to all classifiers: w1=1, w2=1, w3=1.
```
maybe it would be better to have a different number of classes ;)

amueller commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -0,0 +1,49 @@
+
+
+
+if __name__ == "__main__":
```
Drop these, you can just ``nosetests -sv yourfile.py``

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+from ..preprocessing import LabelEncoder
+import numpy as np
+import operator
+
```
two empty lines before classes

amueller commented:
Not sure about using GaussianNB in the examples at that is usually a horrible classifier. Maybe KNN instead?

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+    >>> clf3 = GaussianNB()
+    >>> X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
+    >>> y = np.array([1, 1, 1, 2, 2, 2])
+    >>> eclf1 = VotingClassifier(estimators=[clf1, clf2, clf3], voting='hard')
```
It would be nice to have an example where the outputs differ between the options and where the user can understand why they do. That might be tricky, but at least for the weights it might be possible.

rasbt replied:
I toyed a little bit around with this example, but I really couldn't find a small dataset that would fit into the docstring where the individual classifiers do not agree. Maybe, I have to try something really terribly bad, like KNN with k=1 ... I will keep looking ...

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+            
+    
+    def predict_proba(self, X):
+        """
```
The method should have a description in the first line of the docstring (which can be really stupid)

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,192 @@
+
+  
+    def transform(self, X):
+        """         
```
same comment

amueller commented:
I think the tests could be improved, at least making the tests for the probabilities more explicit (you can create a dataset where you know which mistake will be done by which classifier if you choose them simple enough, maybe?)

Maybe you could add a nice example (with a plot) if you have an idea? Maybe some silly 2d data as in the classifier comparison, and then show how the probability estimates are the averages?

My main concern is that the ``predict`` will not be ``np.argma(predict_proba(X, axis=1))`` for the voting.

rasbt commented:
I am currently in the process to add the suggestions bit by bit. One question about an older one by @amueller : 
You suggested to add an error message to the multilabel case that is not implemented, yet. I am a little uncreative here and can't come up with anything better than:
    
    def fit(self, X, y):
        ...
        if isinstance(y, np.ndarray) and len(y.shape) > 1 and y.shape[1] > 1:
            raise NotImplementedError("Multilabel classification not implemented, yet.")

Any better suggestions?

rasbt commented:
> Maybe you could add a nice example (with a plot) if you have an idea? Maybe some silly 2d data as in the classifier comparison, and then show how the probability estimates are the averages?

Something like this?

![unknown](https://cloud.githubusercontent.com/assets/5618407/6263579/9888e168-b7e7-11e4-8e9f-7cb36c5683bf.png)

rasbt commented:
I think I addressed all of the comments above :) 
Except for the ` GaussianNB` suggestion. In general, I agree, it's not a good classifier, however, I think for those toy examples it might be okay!? I just wanted to have 3 classifiers that have roughly the best performance but don't do too well either for the sake of showing that the VotingClassifier could have an effect.

Hm, now I just see this merge conflict. I guess it has been some time since I fetched updates from scikit-learn, will have a look at (and hopefully resolve) it on the weekend.

rasbt commented:
Should be all set now except for some long doctests lines, E.g., 

    >>> for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
    ...     scores = cross_validation.cross_val_score(clf, X, y, cv=5, scoring='accuracy')
    ...     print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

For some reason, I can't wrap the initial line to e.g., 

    >>> for clf, label in zip([clf1, clf2, clf3, eclf], 
    ...                                  ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
    ...     scores = cross_validation.cross_val_score(clf, X, y, cv=5, scoring='accuracy')
    ...     print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

without causing test failures. Any tips for how to go about it?

ogrisel commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,189 @@
+        self.classes_ = self.le_.classes_
+
+        for clf in self.estimators:
+            clf.fit(X, self.le_.transform(y))      
```
Calling fit on a model should never mutate the constructor parameter state by default. I think we should rewrite this as:

```
self.estimators_ = []
for clf in self.estimators:
    fitted_clf = clone(clf).fit(X, self.le_.transform(y)) 
    self.estimators.append(fitted_clf)
```

Then in the  `_predict` and `_predict_proba` we should use `self.estimators_` instead of `self.estimators`. The newly introduced `estimators_` attribute should be documented in the class docstring.

Also add a `check_fitted(self, 'estimators_')` in the `_predict` and `_predict_proba` methods.

@amueller do you think we should try to write a `test_common.py` to detect those kinds of problems?

ogrisel replied:
> @amueller do you think we should try to write a test_common.py to detect those kinds of problems?

Their is already `check_estimators_overwrite_params`, I will try to upgrade it to detect any mutation.

ogrisel replied:
I created #4274 but this won't be able to cases such as this one unfortunately.

ogrisel commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -0,0 +1,144 @@
+    assert(clf1.fit(X,y).predict(X)[73] == 2) 
+    assert(clf2.fit(X,y).predict(X)[73] == 1) 
+    assert(eclf.fit(X,y).predict(X)[73] == 1) 
+
```
please run the pep8 linter on your new files. In particular here there should be 2 blank lines between top level blocks.

ogrisel commented on sklearn/ensemble/tests/test_voting_classifier.py:
```diff
@@ -0,0 +1,144 @@
+    eclf = VotingClassifier(estimators=[clf1, clf2], voting='hard')
+    assert(clf1.fit(X,y).predict(X)[73] == 2) 
+    assert(clf2.fit(X,y).predict(X)[73] == 1) 
+    assert(eclf.fit(X,y).predict(X)[73] == 1) 
```
Please use `assert_equals` to get informative failure messages.

ogrisel commented on doc/modules/ensemble.rst:
```diff
@@ -893,3 +893,152 @@ averaged.
+
+
+
+
```
2 blank lines is enough to separate blocks.

ogrisel commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,189 @@
+    Parameters
+    ----------
+    estimators : array-like, shape = [n_classifiers]
+      A list of estimators for classification.
```
I would say ".. unfitted estimators ...".

ogrisel commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,189 @@
+    >>>
+    """
+    
+    def __init__(self, estimators, voting='hard', weights=None):
```
Most sklearn estimators should have their constructors with reasonable default values. However I do not see what we could use as default value for estimators in this case. Maybe a we could set `estimators=None` and make the `fit` function treat this case as `[LogisticRegression()]`.

What do people think?

ogrisel replied:
Maybe `[LogisticRegression(), RandomForestClassifier()]` is a reasonable default as they are both classifiers with reasonably well calibrated `predict_proba` and likely to have significantly uncorrelated errors that should often benefit from ensembling.

amueller replied:
I don't think this needs a default. Having meta-estimators is fine. ``GridSearchCV`` and ``Pipeline`` are not default-constructible, and neither is ``RFE``. BaggingClassifier on the other hand has a tree...

ogrisel commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,189 @@
+    voting : str, {'hard', 'soft'} (default='hard')
+      If 'hard', uses predicted class labels for majority rule voting.
+      Else if 'soft', predicts the class label based on the argmax of
+      the sums of the predicted probalities.
```
I would add that `'soft'` is recommended if all the base `estimators` are expected to yield well calibrated probabilities. Once the the calibration PR #1176 is merged we should add a `See also` section to `CalibratedClassifierCV` at the end of the this docstring.

rasbt commented:
@ogrisel I like the idea of having constructors with reasonable default values. However, I am not sure if this is really a good idea for a this `VotingClassifier` since it really depends on your dataset. E.g., throwing in a linear classifier like logistic regression for a non-linear dataset might cause more harm than good in contrast to using e.g., a `RandomForest` alone. I think by not providing a default value here, this would encourage the user to think more about which estimators would make sense to be combined. I think the typical use case would be more something like: I have trained and tuned a bunch of great models, could a "combination" make them even stronger?

rasbt commented:
I am looking forward to discuss the "default values"  further, what do other's think? Besides that, everything that you suggested @ogrisel should be fixed now (I hope), thanks!

amueller commented:
On 02/19/2015 01:18 AM, Sebastian Raschka wrote:
> |def fit(self, X, y):
>      ...
>      if isinstance(y, np.ndarray) and len(y.shape) > 1 and y.shape[1] > 1:
>          raise NotImplementedError("Multilabel classification not implemented, yet.")
> |
>
> Any better suggestions?
>
"Multilabel and multi-output classification is not supported"?

amueller commented:
On 02/19/2015 03:30 AM, Sebastian Raschka wrote:
>
>     Maybe you could add a nice example (with a plot) if you have an
>     idea? Maybe some silly 2d data as in the classifier comparison,
>     and then show how the probability estimates are the averages?
>
> Something like this?
>
>
I thought more about plotting the decision function over the 2D space 
(contourf of decision function), so that you see that the mistakes are 
averaged out.
You could blend a tree and a linear model or a linear model and an rbf 
svm or something to make the two different shapes obvious.

amueller commented:
On 02/19/2015 10:16 PM, Sebastian Raschka wrote:
>
> Should be all set now except for some long doctests lines, E.g.,
>
> |>>> for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
> ...     scores = cross_validation.cross_val_score(clf, X, y, cv=5, scoring='accuracy')
> ...     print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))
> |
>
> For some reason, I can't wrap the initial line to e.g.,
>
> |>>> for clf, label in zip([clf1, clf2, clf3, eclf],
> ...                                  ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
> ...     scores = cross_validation.cross_val_score(clf, X, y, cv=5, scoring='accuracy')
> ...     print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))
> |
>
> without causing test failures. Any tips for how to go about it?
>
>
#doctest: +NORMALIZE_WHITESPACE

amueller commented:
Btw, from an API standpoint I think this should support grid-searching the parameters of the estimators that are averaged. That is done in a slightly different manner in BaggingClassifier as in Pipeline, but I think we should go the pipeline way. Also see #4291.
Currently the easiest option would be to overload ``get_params`` the same way as done in ``Pipeline``. That would require naming the estimators. I propose we just use the estimator naming as introduced by ``make_pipeline`` and don't allow custom strings, as they are pretty useless.

amueller commented:
If you don't feel like doing the grid-search support in this PR, I don't feel it is required, though.

rasbt commented:
> "Multilabel and multi-output classification is not supported"?

I am afraid not, I was just happy to get it to run for the "simple" case ... but a good thing for the future to-do list

> I thought more about plotting the decision function over the 2D space
(contourf of decision function), so that you see that the mistakes are
averaged out.

Ahhhh, yes, something like this?:

![unknown](https://cloud.githubusercontent.com/assets/5618407/6364099/9858b560-bc6f-11e4-9e38-c45ec7d8e40c.png)

> #doctest: +NORMALIZE_WHITESPACE

Hm, somehow I can't get it to work inside function calls ... I will take a deeper look into it on the weekend ...

Btw an option for gridsearching parameters for this sounds very, very useful! I will take a look at this!

amueller commented:
"Multilabel and multi-output classification is not supported"?

>I am afraid not, I was just happy to get it to run for the "simple" case ... but a good thing for the future to-do list

You here asking on a formulation for the error message. sorry if it was not clear from the context that this was my suggestion, not a question on what you implemented.

amueller commented:
And yes, this was the kind of figure I had in mind. thought I am not convinced that particular example you showed is very enlightening ;) But if you don't want to spend time on this, don't worry about it too much. I just thought it might give a useful illustration.

rasbt commented:
@amueller No, I agree with you! I was just a little bit uncreative coming up with good examples! Such a figure is probably a typical example of " ... explaining more than thousands words". I will add it as an example in the next few days when there is more time. Thanks!

rasbt commented:
@amueller I think that GridSearch support would be a good thing for API consistency and also good be useful in practice. However, I have to say that I am struggling with the implementation. Just to make sure that we mean the same thing. You were suggesting something like the following, right?



    from sklearn.preprocessing import StandardScaler
    from sklearn.grid_search import GridSearchCV
    from sklearn.linear_model import LogisticRegression
    from sklearn.naive_bayes import GaussianNB 
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.ensemble import VotingClassifier
    from sklearn import datasets

    iris = datasets.load_iris()

    clf1 = LogisticRegression()
    clf2 = RandomForestClassifier()
    clf3 = GaussianNB()
    eclf = VotingClassifier(estimators=[clf1, clf2, clf3], voting='soft')

    params = {'estimators__clf1__C': [1.0, 10.0, 100.0]}

    grid = GridSearchCV(estimator=eclf, param_grid=params)
    grid.fit(iris.data, iris.target)


Or rather 

        params = {'estimators__logisticregression__C': [1.0, 10.0, 100.0]}

based on `_name_estimators()`. However, for some reason I don't get the concept and implementation right ...


And ideally, it should also  support `pipeline`

    pipeline = Pipeline([('scl', StandardScaler()),
                     ('clf', eclf)])

    param_grid = [{'clf__estimators__logisticregression__C': [0.1, 1, 10, 100],}]

    grid = GridSearchCV(pipeline, 
                           param_grid=param_grid, 
                           verbose=1, 
                           cv=5, 
                           scoring='accuracy', 
                           n_jobs=1)

    grid.fit(iris.data, iris.target)


One question though, what if a user wants to use multiple e.g., logistic regression models, a mix of pre-trained ones and some for `GridSearch` wouldn't this raise conflicts?
Anyway, I am not sure if I am currently able to implement the GridSearch without any further help, sorry!

amueller commented:
Yes, I think we are on the same page about the goal. I would remove the "estimators__" part (pipeline doesn't have it and it doesn't really add anything).

The naming in ``_name_estimators`` uses numbering if you have the same model twice.

I'm not sure what you mean by the pipeline comment. Building more complex pipelines, each step needs to do the delegation for ``set_params`` and ``get_params`` correctly, then everything works out. ``Pipeline`` has that done, so it works with grid-search. If you do it for ``VotingClassifier`` then it will also work for all combinations of pipeline and ``VotingClassifier``.

I'm not sue if I understand your concerns or answered them. Anyhow, I don't feel that this is blocking for the PR.

rasbt commented:
"`estimators__`" is actually not part of the pipeline but the parameter of the VotingClassifier that takes the list of "estimators" .

E.g., 

    params = {'estimators__logisticregression__C': [0.1, 1, 10, 100], 
    'voting': ['soft', 'hard']}, {'weights': [[1,2,3], [2,1,2],}

Sorry about the confusion regarding the `pipeline`:  
> If you do it for VotingClassifier then it will also work for all combinations of pipeline and VotingClassifier.

This was actually what I suspected, but I just wanted to make sure :)

But in general, I am a little bit lost here in terms of what exactly needs to be implemented in the VotingClassifier so that it support GridSearch

amueller commented:
I know the ``estimators__`` is not part of the pipeline, I just think it adds nothing, only makes parameter names longer, and so should be omitted.

What you need to do is overload ``get_params`` similar to how it is done in ``Pipeline``, so that it is recursively called for all estimators in ``estimators``. But as I said, don't sweat it.

rasbt commented:
Ok, I think I got it. I just gave it another try and could figure out what went wrong the last time when I tried the `get_params` overloading: `_name_estimators()` returned a list in my case rather than a dictionary. Anyway, it looks like the `VotingClassifier` has `GridSearch` support now!

I have to say that this little implementation took me way longer than I expected, but, hey, I learned a lot about the scikit-learn internals -- that was a pretty useful exercise!


Notation-wise, I went with your suggestion:

    params = {'logisticregression__C': [1.0, 100.0], 
                     'randomforestclassifier__n_estimators': [20, 200],}
    grid = GridSearchCV(estimator=eclf, param_grid=params, cv=5)
    grid = grid.fit(iris.data, iris.target)

amueller commented:
Sweet! Sorry, contributing something new to sklearn is always a bit of a handful ;)

glouppe commented:
Regarding the auto-naming of classifiers, I would personally prefer to mimic the Pipeline or FeatureUnion API and allow the user to provide a list of `(name, estimator)` tuples. In case a list of estimators is directly provided (as currently done here), then we would resort to default naming. What do  you think @amueller?

glouppe commented:
> I have to say that this little implementation took me way longer than I expected, but, hey, I learned a lot about the scikit-learn internals -- that was a pretty useful exercise!

Thanks for your patience! But indeed, proposing new estimators is really a long process :)

amueller commented:
I'm on the fence about the naming of estimators. We can try to support both, as you suggested. For ``FeatureUnion`` and ``Pipeline`` @larsmans (iirc) introduced ``make_union`` and ``make_pipeline`` basically because the syntax was awkward.

On the other hand, a unified way would be good.
We could accept both versions here, and possibly also accept both versions in ``Pipeline`` and ``FeatureUnion`` in the future, so we wouldn't need the ``make_`` functions?

ogrisel commented:
For the estimator names, what about using integer indices:

```
voting_classifier.set_params({
    'estimator_0__C': 1,
    'estimator_1__max_depth': 10,
    ...
})
```

The caller would have to know which instance / class is passed in which order to the `VotingClassifier` in order to pass correct parameters but I think we can assume that this is the case.

WDYT?

amueller commented:
@ogrisel that is what ``_name_estimators``, as used in ``make_union`` and ``make_pipeline`` does.
(only it uses dashes, not underscores).

rasbt commented:
Okay, so we have the following 2 scenarios:


    clf1 = LogisticRegression
    clf2 = RandomForestClassifier

### Scenario A)

    VotingClassifier(estimators=[clf1, clf2])

    params = {'logisticregression__C': [1.0, 100.0], 
                    'randomforestclassifier__n_estimators': [20, 200],}

### Scenario B)

    VotingClassifier(estimators=[('my_clf1', clf1), ('my_clf2', clf2)])

    params = {'my_clf1__C': [1.0, 100.0], 
                    'my_clf2__n_estimators': [20, 200],}


I like A (that's what's currently implemented) as well as B, and I don't really have a preference here -- but this is also not my decision to make, you are the guys taking care of the API ;). The question is, "A" or "B", or "A and B"? I think that supporting both could make the documentation and examples a little bit confusing for the typical user, and I would suggest to support either "A" or "B" but not both (not only for the `VotingClassifier` but also in general).

amueller commented:
B is what is currently implemented in ``Pipeline`` ... it just feels so verbose... Glad you think someone is "taking care of the API" ;)
I'm a bit torn here. Supporting both might indeed be a bit weird to new users.

glouppe commented:
For now, even if it is verbose, I would be in favour of consistency and therefore would prefer stick to what use for Pipeline and FeatureUnion. 

Then however, I would leave open the possibility to pass a list of estimators (i.e., non-named estimators),  and would implement the same behaviour in Pipeline and FeatureUnion. What do you think? CC: @GaelVaroquaux @larsmans @ogrisel @jnothman

rasbt commented:
Just saw that the release 0.16 is official now. That's awesome, big congrats and thanks for all the good work that you all put into scikit-learn! :). This also reminded me of this open discussion here: Any new thoughts and opinions how we should go about the naming of the estimators?

amueller commented:
Yeah I think we should move on with this. we can use the "named" API for now to mirror Pipeline, and possibly add the other one later

rasbt commented:
Alright, the `VotingClassifier` has the more "traditional" named API now, e.g.,

    clf1 = LogisticRegression(random_state=1)
    clf2 = RandomForestClassifier(random_state=1)

    eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2)])
    params = {'lr__C': [1.0, 100.0],
              'rf__n_estimators': [20, 200],}

    grid = GridSearchCV(estimator=eclf, param_grid=params, cv=5)

Except for this Travis fail... Any thoughts on this? This should be the most current version from the scikit-learn master branch.

<pre>======================================================================
FAIL: sklearn.feature_selection.tests.test_rfe.test_rfe_features_importance
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python3.4/site-packages/nose-1.3.4-py3.4.egg/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/feature_selection/tests/test_rfe.py", line 90, in test_rfe_features_importance
    assert_true(sum(diff_support) == len(diff_support))
nose.proxy.AssertionError: False is not true
    'False is not true' = self._formatMessage('False is not true', "%s is not true" % safe_repr(False))
>>  raise self.failureException('False is not true')
    
----------------------------------------------------------------------
Ran 5027 tests in 192.558s
FAILED (SKIP=16, failures=1) </pre>

amueller commented on sklearn/feature_selection/rfe.py:
```diff
@@ -424,4 +424,4 @@ def fit(self, X, y):
         # here, the scores are normalized by len(cv)
         self.grid_scores_ = scores / len(cv)
-        return self
+        return self
```
no newline at end of file?

amueller commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -254,4 +254,4 @@ def test_rfe_min_step():
     selector = RFE(estimator, step=5)
     sel = selector.fit(X,y)
-    assert_equal(sel.support_.sum(), n_features // 2)
+    assert_equal(sel.support_.sum(), n_features // 2)
```
no newline at end of file?

amueller commented:
The test failure is weird and seems unrelated. Apart from the fact that you touched the rfe files. Can you rebase on master and undo your changes?

rasbt commented:
Well, brushing up my git skills is definitely another item for my to do list.

Anyway,  I think the problem could come from the [Fix RFE](https://github.com/rasbt/scikit-learn/commit/8853dc005b8af5b40e0ba288510d88fc394a046f) commit. But I am really curious why it only causes this issue here and not in the scikit-master branch.

rasbt commented:
Running nose locally also seems to be fine

<pre>~/github/scikit-learn >nosetests sklearn/feature_selection/tests/test_rfe.py -sv
I: Seeding RNGs with 1166611526
sklearn.feature_selection.tests.test_rfe.test_rfe_set_params ... /Users/sebastian/github/scikit-learn/sklearn/feature_selection/rfe.py:145: DeprecationWarning: The parameter 'estimator_params' is deprecated as of version 0.16 and will be removed in 0.18. The parameter is no longer necessary because the value is set via the estimator initialisation or set_params method.
  "set_params method.", DeprecationWarning)
ok
sklearn.feature_selection.tests.test_rfe.test_rfe_features_importance ... ok
sklearn.feature_selection.tests.test_rfe.test_rfe_deprecation_estimator_params ... ok
sklearn.feature_selection.tests.test_rfe.test_rfe ... ok
sklearn.feature_selection.tests.test_rfe.test_rfe_mockclassifier ... ok
sklearn.feature_selection.tests.test_rfe.test_rfecv ... ok
sklearn.feature_selection.tests.test_rfe.test_rfecv_mockclassifier ... ok
sklearn.feature_selection.tests.test_rfe.test_rfe_min_step ... /Users/sebastian/github/scikit-learn/sklearn/feature_selection/rfe.py:184: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  support_[features[ranks][:threshold]] = False
ok

----------------------------------------------------------------------
Ran 8 tests in 2.702s</pre>

amueller commented:
interesting. But all travis versions seem to fail. can you run nosetests on the full folder?

amueller commented:
never mind, it is a missing random state setting.

amueller commented:
your rebase has some issues.... make sure you pull master before doing it.

amueller commented:
culprit: https://github.com/scikit-learn/scikit-learn/pull/4496

ogrisel commented:
> your rebase has some issues.... make sure you pull master before doing it.

=> pull `upstream master` to your local `master` not to this branch (we don't want a merge commit in your feature branch). Alternatively, if you don't want to maintain a local master branch:

```
$ git fetch upstream master
$ git rebase upstream/master
```

rasbt commented:
Sorry & thanks @ogrisel , I really messed up this topic branch, hopefully it looks better now.

amueller commented:
Splendid!

amueller commented:
Failure is fixed here: #4528.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,217 @@
+    >>>
+    """
+
+    def __init__(self, estimators, voting='hard', weights=None):
```
why do we default to hard voting again? Because the classifiers will not be calibrated against each other? [if this is the reason, should we document it?]

rasbt replied:
Hm, good question, I think the decision was eventually pro "hard" to be one the "safe side" since not all classifiers implement `predict_proba` and calibration could also be a potential issue.
 
And maybe this is also a little bit more intuitive (with respect to the term "majority vote") for someone who didn't read the documentation fully. But it is also briefly mentioned in the `VotingClassifier` class doc., and "soft" and "hard" voting have separate sections in the ensemble.rst to explain the difference.

    voting : str, {'hard', 'soft'} (default='hard')
      If 'hard', uses predicted class labels for majority rule voting.
      Else if 'soft', predicts the class label based on the argmax of
      the sums of the predicted probabilities, which is recommended for
      an ensemble of well-calibrated classifiers.

amueller replied:
yeah ok makes sense.

amueller commented:
Sorry for delaying again, but I'm still a bit concerned that ``predict_proba`` is not consistent with ``predict`` for hard voting (which is the default). The ``OvRClassifier`` has an ugly hack to deal with that....
Also, we are currently breaking ties always in favor of the first class, right? Is that good?

amueller commented:
Maybe the tie-breaking is not that big a concern and can be left for a future PR, but the ``predict_proba`` ....

amueller commented:
[ping @jnothman @ogrisel ?]

rasbt commented:
@amueller  

> "but I'm still a bit concerned that `predict_proba` is not consistent with `predict` for hard voting (which is the default)" 

You mean because of not "well calibrated" classifiers?

E.g., if a hard vote majority rule for a binary task would be: 


clf1 -> 0
clf2 -> 0
clf3 -> 1

predict: class 0

the underlying probas could be:

class 1 avg. proba: (0.49 + 0.49 + 0.99) / 3

class 0 avg. proba:  (0.51 + 0.51 + 0.01) / 3

predict: class 1

I am not sure if this is a bad thing. It's just "different" and may be what the user wants or doesn't want.

amueller commented:
the user might count on predict being the argmax of predict_proba. I'm not sure we actually promise this, but is is a bit unintuitive if it is not true.

amueller commented:
@agramfort proposed to raise a ``NotImplementedError`` on ``predict_proba`` when doing hard averaging. I think that's a good idea. what do you think?

agramfort commented:
and add a probability=True or False as param as SVC to define averaging
behavior (hard or soft voting)

mblondel commented:
A solution that makes `hasattr(clf, "predict_proba")` work would be nice.

agramfort commented:
if user sets probability=True or voting=soft it is his responsibility to
assume predict_proba works. We've had this in SVC for years an nobody
complained. Let's be pragmatic here.

rasbt commented:
So, the plan would be to use `predict_proba` (instead of `predict`) of the estimators for `"hard"` voting if a parameter `probability=True` in the `VotingClassifier`? I would favor this slightly over an automatic approach using `hasattr(clf, "predict_proba")` to provide the user with some (optional) more flexibility. I would set the default `probability=False` though; is there really a difference between using `predict_proba` and `predict` for `'hard'` voting?

agramfort commented:
hard would use predict
soft would use predict_proba

rasbt commented:
Oh, okay, so instead of the current averaging approach in the "soft" voting would be:

## a) (current)

E.g., 3 classifiers would currently predict the probabilities that 1 sample belongs to class 0 or 1: 

    class0: (0.45, 0.56, 0.73) / 3 = 0.58
    class1: (0.55, 0.44, 0.27) / 3  = 0.42

    -> predict class 0

and the proposed change for soft voting would then be something like:

## b)


    class0: (0.45, 0.56, 0.73)  -> 0, 1, 1
    class1: (0.55, 0.44, 0.27)  -> 1, 0, 0

    -> predict class 0


When @amueller suggested

> the user might count on predict being the argmax of predict_proba. I'm not sure we actually promise this, but is is a bit unintuitive if it is not true.

I was thinking of something else. Now, I agree that the results between the currently implemented "soft" and "hard" voting can differ in certain situation -- I thought that it could be useful to have this option. If we change the implementation of "soft" to the scenario b), the hard and soft voting would be identical --- wouldn't "soft" voting become redundant in this case?

amueller commented:
I am now confused.
I think it is good to have alternative hard and soft voting strategies, and I don't suggest any change to them. I think the idea was that when ``voting="soft"`` we don't implement ``predict_proba`` because there is no way to do that consistently.

rasbt commented:
@amueller Okay, I guess I confused the `VotingClassifier.predict_proba` with the individual estimators'  `predict_proba` when I read the suggestions.

Hopefully, I am on the same page now, so this would simply come down to:

    def predict_proba(self, X):
       ...
        if self.voting == 'hard':
            raise NotImplementedError("predict_proba is only supported in 'soft' voting.")
        
        avg = np.average(self._predict_probas(X), axis=0, weights=self.weights)
        return avg

amueller commented:
that is what I had in mind.

rasbt commented:
Okay, good! Then the NotImplementedError should be implemented now.

rasbt commented:
Sorry for the additional rebase, I somehow lost the overview over my local fork and there was this path for the random_state in RFE that I wanted pull for the Travis build

amueller commented:
you should actually rebase, not merge.
Also, maybe squash the commits.
Do you want to add an entry in the whatsnew.rst (under 0.17 new features)?

@mblondel can you live with this solution?

rasbt commented:
@amueller okay, let me see if I can fix the previous mess and then I will try the squashing

rasbt committed:
votingclassifier

mblondel commented:
I would rather use the same technique as for SVC
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/svm/base.py#L547

I actually don't agree with @larsmans 's inline comment: introducing two estimators just for that sounds kind of silly.

larsmans commented:
I still think that's the way to keep the codebase readable and maintainable. The `SVC.predict_proba` trick changes the type of an object after it's constructed, and it means that objects may not (appear to) have a method that the class does have.

I do prefer it to a `NotImplementedError`, though; that would be yet another convention to check for in meta-estimators.

rasbt commented:
Okay, so we ignore the part 

> Do not use this in new code; when
    # probabilities are not available depending on a setting, introduce two
    # estimators.

for now?
 I am just wondering if we need the `probability` parameter. Can't we just check `if self.voting=='hard'`;

instead of

        if not self.probability:
            raise AttributeError("predict_proba is not available when"
                             " probability=%r" % self.probability)

agramfort commented:
seems legit..

rasbt committed:
added _check_proba method

amueller commented:
yeah I don't think we need a probability parameter, and I agree with using the ``SVC`` trick. It seems quite more harmless than the now abundant ``if_delegate_has_method``.

rasbt commented:
Have to admit that I don't really see through this "property" trick. How would the `@property` decorator come into play here if we simply check whether the VotingClassifier was initialized for hard or soft voting.

amueller commented:
the code inside the @property should raise an ``AttributeError`` if  ``voting="hard"`` and should return the the function that actually implements ``predict_proba`` otherwise.

A property triggers a function evaluation, but looks like an attribute access. So trying to access the ``predict_proba`` function will actually run the function.

rasbt committed:
property trick to raise AttributeError if voting is "hard"

rasbt commented:
Thanks. I hope the most recent commit implements what you were looking for.

rasbt commented:
Before this pull request sets on some dust again, I was wondering if everything seems to be fine now!?

amueller commented:
+1. Thanks for the ping.

amueller commented on sklearn/ensemble/voting_classifier.py:
```diff
@@ -0,0 +1,227 @@
+        avg : array-like, shape = [n_samples, n_classes]
+            Weighted average probability for each class per sample.
+        """
+        self._check_proba()
```
You could inline this (the function is only called once) but... well.. however you like..

rasbt replied:
Hm, yes, I also initially thought that this could be a little bit of an overkill ... but maybe it's slightly better this way in case there will be a more general check_proba function in some distant future

amueller replied:
[cough](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) but really I don't have a strong opinion ;)

rasbt replied:
Okay, okay, that's strong enough :)

rasbt committed:
inline check_proba

glouppe commented:
I would have some cosmetic changes to do, but nothing significant. +1 for merge before it stalls.

agramfort commented:
thanks @rasbt !

amueller commented:
@glouppe what changes? Do you want to open an issue or open a PR?

rasbt commented:
Just saw that it finally got merged! :) Thank you all for the great help during the implementation. Since I am a big fan and frequent user of scikit-learn, it's really nice contribute, and I really learned a few great lessons here!

glouppe commented:
Thanks for your contribution!

@amueller  I did make some  changes already  in a following commit. Only
cosmetics though
Le 8 mai 2015 20:13, "Sebastian Raschka" <notifications@github.com> a
écrit :

> Just saw that it finally got merged! :) Thank you all for the great help
> during the implementation. Since I am a big fan and frequent user of
> scikit-learn, it's really nice contribute, and I really learned a few great
> lessons here!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4161#issuecomment-100317829>
> .
>

amueller commented:
@rasbt thanks for the contribution and your patience!

vene commented on sklearn/utils/testing.py:
```diff
@@ -505,12 +505,14 @@ def uninstall_mldata_mock():
 def all_estimators(include_meta_estimators=False,
                    include_other=False, type_filter=None,
                    include_dont_test=False):
+                   
```
The Case of the Mysterious Blank Lines?

jnothman commented:
It looks to me like the intro to ensembles (top of http://scikit-learn.org/dev/modules/ensemble.html) should be updated

rasbt commented:
@jnothman I agree. Although the VotingClassifier approach is different from bagging, I'd still but the former into the "averaging methods" category.
